File: backend/src/services/menuItem.service.ts
import { categoryProvider } from '../providers/categoryProvider.js';
import { menuItemProvider } from '../providers/menuItemProvider.js';
import { menuItemImageProvider } from '../providers/menuItemImageProvider.js';
import {
  MenuItemCreateBodyType,
  MenuItemDetailResType,
  MenuItemRes,
  MenuItemUpdateBodyType,
} from '../schemas/menuItem.js';
import { BaseListResType, BaseSearchRequestType } from '../schemas/search.js';
import { AppError } from '../utils/AppError.js';
import { staticFileService } from './staticFile.service.js';

/**
 * Service Layer — MenuItem
 * Maps raw Prisma rows to the API response shape.
 */
export const menuItemService = {
  async getAll(request: BaseSearchRequestType): Promise<BaseListResType> {
    const limit = request.limit;
    const rows = await menuItemProvider.findAll(request);

    const hasNextPage = rows.length > limit;
    const items = hasNextPage ? rows.slice(0, limit) : rows;
    const nextCursor = hasNextPage ? items[items.length - 1].id : null;
    const normalize = (r: any) => {
      const price =
        r.price &&
        typeof r.price === 'object' &&
        typeof r.price.toNumber === 'function'
          ? r.price.toNumber()
          : r.price != null
            ? Number(r.price)
            : r.price;
      const rating =
        r.rating != null &&
        typeof r.rating === 'object' &&
        typeof r.rating.toNumber === 'function'
          ? r.rating.toNumber()
          : r.rating != null
            ? Number(r.rating)
            : r.rating;

      return {
        ...r,
        price,
        rating,
        commentCount: r._count?.menu_item_comments ?? 0,
        image: r.menu_item_images?.[0]?.image_url ?? null,
        created: r.created instanceof Date ? r.created : new Date(r.created),
        updated: r.updated
          ? r.updated instanceof Date
            ? r.updated
            : new Date(r.updated)
          : null,
      };
    };

    return {
      data: items.map((c) => MenuItemRes.parse(normalize(c))),
      limit: request.limit,
      nextCursor,
      hasNextPage,
    };
  },

  async findById(id: number): Promise<MenuItemDetailResType | null> {
    const row = await menuItemProvider.findById(id);
    if (!row) return null;

    const r = row as any;

    const toNum = (v: any) =>
      v != null && typeof v === 'object' && typeof v.toNumber === 'function'
        ? v.toNumber()
        : v != null
          ? Number(v)
          : null;

    const imgs: any[] = r.menu_item_images ?? [];

    return {
      id: r.id.toString(),
      name: r.name,
      description: r.description ?? null,
      price: toNum(r.price) ?? 0,
      rating: r.rating != null ? toNum(r.rating) : null,
      commentCount: r._count?.menu_item_comments ?? 0,
      tag: r.tag ?? null,
      category: r.category
        ? { id: r.category.id.toString(), name: r.category.name }
        : null,
      images: imgs.map((img) => ({
        id: img.id.toString(),
        image_url: img.image_url,
        is_primary: img.is_primary ?? false,
        display_order: img.display_order ?? 0,
        created_at: img.created_at.toISOString(),
      })),
      created: r.created.toISOString(),
      updated: r.updated ? r.updated.toISOString() : null,
      deleted: r.deleted,
      active: r.active,
    };
  },

  async uploadImages(
    menuItemId: number,
    files: Express.Multer.File[],
    isPrimary?: string,
  ): Promise<MenuItemDetailResType | null> {
    const item = await menuItemProvider.findById(menuItemId);
    if (!item) throw new AppError(404, 'Không tìm thấy món ăn');

    const primaryIdx = isPrimary !== undefined ? Number(isPrimary) : null;
    const images = files.map((file, idx) => ({
      image_url: staticFileService.getPath(file),
      is_primary: primaryIdx !== null ? idx === primaryIdx : idx === 0,
      display_order: idx,
    }));

    await menuItemImageProvider.createMany(menuItemId, images);
    return this.findById(menuItemId);
  },

  async deleteImage(
    menuItemId: number,
    imageId: number,
  ): Promise<MenuItemDetailResType | null> {
    const img = await menuItemImageProvider.findById(imageId);
    if (!img) throw new AppError(404, 'Không tìm thấy ảnh');
    if (img.menu_item_id.toString() !== menuItemId.toString())
      throw new AppError(400, 'Ảnh không thuộc món ăn này');

    // Delete file from disk
    try {
      staticFileService.delete(img.image_url);
    } catch {
      /* ignore */
    }

    await menuItemImageProvider.delete(imageId);
    return this.findById(menuItemId);
  },

  async create(data: MenuItemCreateBodyType) {
    const cate = await categoryProvider.findById(data.category_id);
    if (!cate) throw new AppError(404, 'Không tìm thấy loại món ăn');

    return menuItemProvider.create(data);
  },

  async update(data: MenuItemUpdateBodyType, id: number) {
    return menuItemProvider.update(data, id);
  },

  async delete(id: number) {
    return menuItemProvider.delete(id);
  },
};

File: backend/app/services/menu_item_service.py
from sqlalchemy.ext.asyncio import AsyncSession
from app.providers.menu_item_provider import menu_item_provider
from app.providers.menu_item_image_provider import menu_item_image_provider
from app.utils.app_error import AppError
from typing import Optional

async def get_all(db: AsyncSession, limit: int = 10, cursor: Optional[int] = None, order: str = "desc", search: Optional[str] = None, category_id: Optional[int] = None) -> dict:
    rows = await menu_item_provider.find_all(db, limit=limit, cursor=cursor, order=order, search=search, category_id=category_id)
    has_next = len(rows) > limit
    items = rows[:limit] if has_next else rows
    next_cursor = items[-1].id if has_next and items else None
    return {"data": items, "limit": limit, "nextCursor": next_cursor, "hasNextPage": has_next}

async def find_by_id(db: AsyncSession, item_id: int):
    item = await menu_item_provider.find_by_id(db, item_id)
    if not item:
        raise AppError(404, "Menu item not found")
    return item

async def create(db: AsyncSession, dto: dict):
    data = {k: v for k, v in dto.items() if v is not None and k != "image"}
    return await menu_item_provider.create(db, **data)

async def update(db: AsyncSession, item_id: int, dto: dict):
    data = {k: v for k, v in dto.items() if v is not None and k != "image"}
    item = await menu_item_provider.update(db, item_id, **data)
    if not item:
        raise AppError(404, "Menu item not found")
    return item

async def delete(db: AsyncSession, item_id: int):
    item = await menu_item_provider.soft_delete(db, item_id)
    if not item:
        raise AppError(404, "Menu item not found")
    return item

async def add_image(db: AsyncSession, menu_item_id: int, image_url: str, is_primary: bool = False) -> dict:
    return await menu_item_image_provider.create(db, menu_item_id, image_url, is_primary)

async def delete_image(db: AsyncSession, image_id: int) -> None:
    await menu_item_image_provider.delete(db, image_id)

async def set_primary_image(db: AsyncSession, menu_item_id: int, image_id: int) -> None:
    await menu_item_image_provider.set_primary(db, menu_item_id, image_id)
